// @aws-cdk/aws-certificatemanager/lib/certificate-base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertificateBase = void 0;
const cloudwatch = require("@aws-cdk/aws-cloudwatch");
const aws_cloudwatch_1 = require("@aws-cdk/aws-cloudwatch");
const core_1 = require("@aws-cdk/core");
/**
 * Shared implementation details of ICertificate implementations.
 *
 * @internal
 */
class CertificateBase extends core_1.Resource {
    /**
     * Return the DaysToExpiry metric for this AWS Certificate Manager Certificate. By default, this is the minimum value over 1 day.
     *
     * This metric is no longer emitted once the certificate has effectively
     * expired, so alarms configured on this metric should probably treat missing
     * data as "breaching".
     *
     * @stability stable
     */
    metricDaysToExpiry(props) {
        return new cloudwatch.Metric({
            period: core_1.Duration.days(1),
            ...props,
            dimensions: { CertificateArn: this.certificateArn },
            metricName: 'DaysToExpiry',
            namespace: 'AWS/CertificateManager',
            region: this.region,
            statistic: aws_cloudwatch_1.Statistic.MINIMUM,
        });
    }
}
exports.CertificateBase = CertificateBase;
